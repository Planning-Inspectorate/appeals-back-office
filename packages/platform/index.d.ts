import { HandlerFunction } from 'got';

declare module '@pins/platform' {
	export function createTtlHandler(): HandlerFunction;
	export function validatePostcode(value: string): boolean;
	export function validatePastDate(value: Date | string | number): boolean;
	export function validateFutureDate(value: Date | string | number): boolean;
	export function yesterday(): Date;
}

export interface LocalPlanningDepartment {
	FID: number;
	LPA21CD: string;
	LPA21NM: string;
	Co_terminous: 'Y' | 'N';
}
