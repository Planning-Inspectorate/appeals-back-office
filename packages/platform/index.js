export * from './http/ttl.js';
export * from './util/date.js';
export * from './validators/date.js';
export * from './validators/postcode.js';
