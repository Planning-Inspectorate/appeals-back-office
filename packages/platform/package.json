{
  "name": "@pins/platform",
  "version": "0.0.0",
  "type": "module",
  "main": "index.js",
  "private": true,
  "engines": {
    "node": ">=22.0.0 <= 22.14.0"
  },
  "scripts": {
    "clean": "rimraf .turbo node_modules dist",
    "lint:js": "npx eslint .",
    "lint:js:fix": "npx eslint . --fix",
    "test": "cross-env NODE_ENV=test NODE_OPTIONS=\"--experimental-vm-modules --expose-gc\" npx jest --logHeapUsage --ci",
    "test:cov": "cross-env NODE_ENV=test NODE_OPTIONS=\"--experimental-vm-modules --expose-gc\" npx jest --logHeapUsage --ci --coverage",
    "tscheck": "npx tsc -p jsconfig.json --maxNodeModuleJsDepth 0"
  },
  "dependencies": {
    "date-fns": "2.28.0",
    "dotenv": "16.0.3",
    "got": "^12.5.2",
    "html": "^1.0.0",
    "lodash-es": "^4.17.21",
    "node-html-parser": "^5.3.3",
    "randexp": "^0.5.3"
  },
  "devDependencies": {
    "@types/html": "^1.0.4",
    "@types/lodash-es": "^4.17.12",
    "cross-env": "^7.0.3",
    "jest": "^29.7.0",
    "type-fest": "^4.37.0"
  }
}
