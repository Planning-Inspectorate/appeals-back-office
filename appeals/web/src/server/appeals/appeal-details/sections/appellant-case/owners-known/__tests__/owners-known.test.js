// @ts-nocheck
import { parseHtml } from '@pins/platform';
import supertest from 'supertest';
import { appealData, appellantCaseDataNotValidated } from '#testing/app/fixtures/referencedata.js';
import { createTestEnvironment } from '#testing/index.js';
import nock from 'nock';

const { app, installMockApi, teardown } = createTestEnvironment();
const request = supertest(app);
const baseUrl = '/appeals-service/appeal-details';

describe('owners-known', () => {
	beforeEach(installMockApi);
	afterEach(teardown);

	describe('GET /change', () => {
		it('should render the change owners known page with "No" radio option checked if knowsOtherLandowners is "No"', async () => {
			nock('http://test/')
				.get(`/appeals/1/appellant-cases/${appealData.appellantCaseId}`)
				.reply(200, {
					...appellantCaseDataNotValidated,
					siteOwnership: {
						...appellantCaseDataNotValidated.siteOwnership,
						knowsOtherLandowners: 'No'
					}
				});

			const response = await request.get(`${baseUrl}/1/appellant-case/owners-known/change`);
			const element = parseHtml(response.text);

			expect(element.innerHTML).toMatchSnapshot();

			const unprettifiedElement = parseHtml(response.text, { skipPrettyPrint: true });

			expect(unprettifiedElement.innerHTML).toContain('Change owners known</h1>');
			expect(unprettifiedElement.innerHTML).toContain(
				'name="ownersKnownRadio" type="radio" value="No" checked>'
			);
			expect(unprettifiedElement.innerHTML).toContain(
				'name="ownersKnownRadio" type="radio" value="Yes">'
			);
			expect(unprettifiedElement.innerHTML).toContain(
				'name="ownersKnownRadio" type="radio" value="Some">'
			);
			expect(unprettifiedElement.innerHTML).toContain(
				'name="ownersKnownRadio" type="radio" value="not-applicable">'
			);
			expect(unprettifiedElement.innerHTML).toContain('Continue</button>');
		});

		it('should render the change owners known page with "Yes" radio option checked if knowsOtherLandowners is "Yes"', async () => {
			nock('http://test/')
				.get(`/appeals/1/appellant-cases/${appealData.appellantCaseId}`)
				.reply(200, {
					...appellantCaseDataNotValidated,
					siteOwnership: {
						...appellantCaseDataNotValidated.siteOwnership,
						knowsOtherLandowners: 'Yes'
					}
				});

			const response = await request.get(`${baseUrl}/1/appellant-case/owners-known/change`);
			const element = parseHtml(response.text);

			expect(element.innerHTML).toMatchSnapshot();

			const unprettifiedElement = parseHtml(response.text, { skipPrettyPrint: true });

			expect(unprettifiedElement.innerHTML).toContain('Change owners known</h1>');
			expect(unprettifiedElement.innerHTML).toContain(
				'name="ownersKnownRadio" type="radio" value="No">'
			);
			expect(unprettifiedElement.innerHTML).toContain(
				'name="ownersKnownRadio" type="radio" value="Yes" checked>'
			);
			expect(unprettifiedElement.innerHTML).toContain(
				'name="ownersKnownRadio" type="radio" value="Some">'
			);
			expect(unprettifiedElement.innerHTML).toContain(
				'name="ownersKnownRadio" type="radio" value="not-applicable">'
			);
			expect(unprettifiedElement.innerHTML).toContain('Continue</button>');
		});

		it('should render the change owners known page with "Some" radio option checked if knowsOtherLandowners is "Some"', async () => {
			nock('http://test/')
				.get(`/appeals/1/appellant-cases/${appealData.appellantCaseId}`)
				.reply(200, {
					...appellantCaseDataNotValidated,
					siteOwnership: {
						...appellantCaseDataNotValidated.siteOwnership,
						knowsOtherLandowners: 'Some'
					}
				});

			const response = await request.get(`${baseUrl}/1/appellant-case/owners-known/change`);
			const element = parseHtml(response.text);

			expect(element.innerHTML).toMatchSnapshot();

			const unprettifiedElement = parseHtml(response.text, { skipPrettyPrint: true });

			expect(unprettifiedElement.innerHTML).toContain('Change owners known</h1>');
			expect(unprettifiedElement.innerHTML).toContain(
				'name="ownersKnownRadio" type="radio" value="No">'
			);
			expect(unprettifiedElement.innerHTML).toContain(
				'name="ownersKnownRadio" type="radio" value="Yes">'
			);
			expect(unprettifiedElement.innerHTML).toContain(
				'name="ownersKnownRadio" type="radio" value="Some" checked>'
			);
			expect(unprettifiedElement.innerHTML).toContain(
				'name="ownersKnownRadio" type="radio" value="not-applicable">'
			);
			expect(unprettifiedElement.innerHTML).toContain('Continue</button>');
		});

		it('should render the change owners known page with "not applicable" radio option checked if knowsOtherLandowners is null', async () => {
			nock('http://test/')
				.get(`/appeals/1/appellant-cases/${appealData.appellantCaseId}`)
				.reply(200, {
					...appellantCaseDataNotValidated,
					siteOwnership: {
						...appellantCaseDataNotValidated.siteOwnership,
						knowsOtherLandowners: null
					}
				});

			const response = await request.get(`${baseUrl}/1/appellant-case/owners-known/change`);
			const element = parseHtml(response.text);

			expect(element.innerHTML).toMatchSnapshot();

			const unprettifiedElement = parseHtml(response.text, { skipPrettyPrint: true });

			expect(unprettifiedElement.innerHTML).toContain('Change owners known</h1>');
			expect(unprettifiedElement.innerHTML).toContain(
				'name="ownersKnownRadio" type="radio" value="No">'
			);
			expect(unprettifiedElement.innerHTML).toContain(
				'name="ownersKnownRadio" type="radio" value="Yes">'
			);
			expect(unprettifiedElement.innerHTML).toContain(
				'name="ownersKnownRadio" type="radio" value="Some">'
			);
			expect(unprettifiedElement.innerHTML).toContain(
				'name="ownersKnownRadio" type="radio" value="not-applicable" checked>'
			);
			expect(unprettifiedElement.innerHTML).toContain('Continue</button>');
		});
	});

	describe('POST /change', () => {
		const validValues = ['No', 'Yes', 'Some', null];

		for (const validValue of validValues) {
			it(`should call appellant cases PATCH endpoint and redirect to the appellant case page if ${
				validValue === null ? validValue : `"${validValue}"`
			} is selected`, async () => {
				const mockAppellantCasesPatchEndpoint = nock('http://test/')
					.patch(`/appeals/1/appellant-cases/${appealData.appellantCaseId}`)
					.reply(200, {});

				const response = await request
					.post(`${baseUrl}/1/appellant-case/owners-known/change`)
					.send({
						knowsOtherOwners: validValue
					});

				expect(mockAppellantCasesPatchEndpoint.isDone()).toBe(true);
				expect(response.statusCode).toBe(302);
				expect(response.text).toBe(
					'Found. Redirecting to /appeals-service/appeal-details/1/appellant-case'
				);
			});
		}
	});
});
