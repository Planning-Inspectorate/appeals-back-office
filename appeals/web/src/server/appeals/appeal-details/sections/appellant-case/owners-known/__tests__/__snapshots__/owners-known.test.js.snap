// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`owners-known GET /change should render the change owners known page with "No" radio option checked if knowsOtherLandowners is "No" 1`] = `
"<main class="govuk-main-wrapper " id="main-content" role="main">
    <div class="govuk-grid-row">
        <div class="govuk-grid-column-two-thirds">
            <div class="govuk-grid-row">
                <div class="govuk-grid-column-full"><span class="govuk-caption-l govuk-!-margin-bottom-3">Appeal 351062</span>
                    <h1                     class="govuk-heading-l govuk-!-margin-bottom-3">Change owners known</h1>
                </div>
            </div>
            <form method="POST" novalidate="novalidate">
                <div class="govuk-form-group">
                    <div class="govuk-radios" data-module="govuk-radios">
                        <div class="govuk-radios__item">
                            <input class="govuk-radios__input" id="ownersKnownRadio" name="ownersKnownRadio"
                            type="radio" value="No" checked>
                            <label class="govuk-label govuk-radios__label" for="ownersKnownRadio">No</label>
                        </div>
                        <div class="govuk-radios__item">
                            <input class="govuk-radios__input" id="ownersKnownRadio-2" name="ownersKnownRadio"
                            type="radio" value="Yes">
                            <label class="govuk-label govuk-radios__label" for="ownersKnownRadio-2">Yes</label>
                        </div>
                        <div class="govuk-radios__item">
                            <input class="govuk-radios__input" id="ownersKnownRadio-3" name="ownersKnownRadio"
                            type="radio" value="Some">
                            <label class="govuk-label govuk-radios__label" for="ownersKnownRadio-3">Some</label>
                        </div>
                        <div class="govuk-radios__item">
                            <input class="govuk-radios__input" id="ownersKnownRadio-4" name="ownersKnownRadio"
                            type="radio" value="not-applicable">
                            <label class="govuk-label govuk-radios__label" for="ownersKnownRadio-4">Not applicable</label>
                        </div>
                    </div>
                </div>
                <button type="submit" data-prevent-double-click="true" class="govuk-button"
                data-module="govuk-button">Continue</button>
            </form>
        </div>
    </div>
</main>"
`;

exports[`owners-known GET /change should render the change owners known page with "Some" radio option checked if knowsOtherLandowners is "Some" 1`] = `
"<main class="govuk-main-wrapper " id="main-content" role="main">
    <div class="govuk-grid-row">
        <div class="govuk-grid-column-two-thirds">
            <div class="govuk-grid-row">
                <div class="govuk-grid-column-full"><span class="govuk-caption-l govuk-!-margin-bottom-3">Appeal 351062</span>
                    <h1                     class="govuk-heading-l govuk-!-margin-bottom-3">Change owners known</h1>
                </div>
            </div>
            <form method="POST" novalidate="novalidate">
                <div class="govuk-form-group">
                    <div class="govuk-radios" data-module="govuk-radios">
                        <div class="govuk-radios__item">
                            <input class="govuk-radios__input" id="ownersKnownRadio" name="ownersKnownRadio"
                            type="radio" value="No">
                            <label class="govuk-label govuk-radios__label" for="ownersKnownRadio">No</label>
                        </div>
                        <div class="govuk-radios__item">
                            <input class="govuk-radios__input" id="ownersKnownRadio-2" name="ownersKnownRadio"
                            type="radio" value="Yes">
                            <label class="govuk-label govuk-radios__label" for="ownersKnownRadio-2">Yes</label>
                        </div>
                        <div class="govuk-radios__item">
                            <input class="govuk-radios__input" id="ownersKnownRadio-3" name="ownersKnownRadio"
                            type="radio" value="Some" checked>
                            <label class="govuk-label govuk-radios__label" for="ownersKnownRadio-3">Some</label>
                        </div>
                        <div class="govuk-radios__item">
                            <input class="govuk-radios__input" id="ownersKnownRadio-4" name="ownersKnownRadio"
                            type="radio" value="not-applicable">
                            <label class="govuk-label govuk-radios__label" for="ownersKnownRadio-4">Not applicable</label>
                        </div>
                    </div>
                </div>
                <button type="submit" data-prevent-double-click="true" class="govuk-button"
                data-module="govuk-button">Continue</button>
            </form>
        </div>
    </div>
</main>"
`;

exports[`owners-known GET /change should render the change owners known page with "Yes" radio option checked if knowsOtherLandowners is "Yes" 1`] = `
"<main class="govuk-main-wrapper " id="main-content" role="main">
    <div class="govuk-grid-row">
        <div class="govuk-grid-column-two-thirds">
            <div class="govuk-grid-row">
                <div class="govuk-grid-column-full"><span class="govuk-caption-l govuk-!-margin-bottom-3">Appeal 351062</span>
                    <h1                     class="govuk-heading-l govuk-!-margin-bottom-3">Change owners known</h1>
                </div>
            </div>
            <form method="POST" novalidate="novalidate">
                <div class="govuk-form-group">
                    <div class="govuk-radios" data-module="govuk-radios">
                        <div class="govuk-radios__item">
                            <input class="govuk-radios__input" id="ownersKnownRadio" name="ownersKnownRadio"
                            type="radio" value="No">
                            <label class="govuk-label govuk-radios__label" for="ownersKnownRadio">No</label>
                        </div>
                        <div class="govuk-radios__item">
                            <input class="govuk-radios__input" id="ownersKnownRadio-2" name="ownersKnownRadio"
                            type="radio" value="Yes" checked>
                            <label class="govuk-label govuk-radios__label" for="ownersKnownRadio-2">Yes</label>
                        </div>
                        <div class="govuk-radios__item">
                            <input class="govuk-radios__input" id="ownersKnownRadio-3" name="ownersKnownRadio"
                            type="radio" value="Some">
                            <label class="govuk-label govuk-radios__label" for="ownersKnownRadio-3">Some</label>
                        </div>
                        <div class="govuk-radios__item">
                            <input class="govuk-radios__input" id="ownersKnownRadio-4" name="ownersKnownRadio"
                            type="radio" value="not-applicable">
                            <label class="govuk-label govuk-radios__label" for="ownersKnownRadio-4">Not applicable</label>
                        </div>
                    </div>
                </div>
                <button type="submit" data-prevent-double-click="true" class="govuk-button"
                data-module="govuk-button">Continue</button>
            </form>
        </div>
    </div>
</main>"
`;

exports[`owners-known GET /change should render the change owners known page with "not applicable" radio option checked if knowsOtherLandowners is null 1`] = `
"<main class="govuk-main-wrapper " id="main-content" role="main">
    <div class="govuk-grid-row">
        <div class="govuk-grid-column-two-thirds">
            <div class="govuk-grid-row">
                <div class="govuk-grid-column-full"><span class="govuk-caption-l govuk-!-margin-bottom-3">Appeal 351062</span>
                    <h1                     class="govuk-heading-l govuk-!-margin-bottom-3">Change owners known</h1>
                </div>
            </div>
            <form method="POST" novalidate="novalidate">
                <div class="govuk-form-group">
                    <div class="govuk-radios" data-module="govuk-radios">
                        <div class="govuk-radios__item">
                            <input class="govuk-radios__input" id="ownersKnownRadio" name="ownersKnownRadio"
                            type="radio" value="No">
                            <label class="govuk-label govuk-radios__label" for="ownersKnownRadio">No</label>
                        </div>
                        <div class="govuk-radios__item">
                            <input class="govuk-radios__input" id="ownersKnownRadio-2" name="ownersKnownRadio"
                            type="radio" value="Yes">
                            <label class="govuk-label govuk-radios__label" for="ownersKnownRadio-2">Yes</label>
                        </div>
                        <div class="govuk-radios__item">
                            <input class="govuk-radios__input" id="ownersKnownRadio-3" name="ownersKnownRadio"
                            type="radio" value="Some">
                            <label class="govuk-label govuk-radios__label" for="ownersKnownRadio-3">Some</label>
                        </div>
                        <div class="govuk-radios__item">
                            <input class="govuk-radios__input" id="ownersKnownRadio-4" name="ownersKnownRadio"
                            type="radio" value="not-applicable" checked>
                            <label class="govuk-label govuk-radios__label" for="ownersKnownRadio-4">Not applicable</label>
                        </div>
                    </div>
                </div>
                <button type="submit" data-prevent-double-click="true" class="govuk-button"
                data-module="govuk-button">Continue</button>
            </form>
        </div>
    </div>
</main>"
`;
